src/Core/Lifecycle.js

```diff
@@ -0,0 +1,78 @@
+/**
+ * @param {Element} element
+ * @param {Object?} state
+ */
+function lifecycleReturn (element, state) {
+	switch (typeof state) {
+		case 'object':
+			if (!state)
+				break
+		case 'function':
+			if (client)
+				enqueueState(element, element.instance, state)
+	}
+}
+
+/**
+ * @param {Element} element
+ * @param {function} callback
+ * @param {*} primary
+ * @param {*} secondary
+ * @param {*} optional
+ */
+function lifecycleCallback (element, callback, primary, secondary, optional) {
+	try {
+		return callback.call(element.instance, primary, secondary, optional)
+	} catch (e) {
+		errorBoundary(element, e, LifecycleCallback, 0)
+	}
+}
+
+/**
+ * @param {Element} element
+ * @param {string} name
+ */
+function lifecycleData (element, name) {
+	try {
+		return element.owner[name].call(element.instance)
+	} catch (e) {
+		errorBoundary(element, e, name, 1)
+	}
+}
+
+/**
+ * @param {Element} element
+ * @param {string} name
+ */
+function lifecycleMount (element, name) {
+	try {
+		var state = element.owner[name].call(element.instance, findDOMNode(element))
+		
+		if (state instanceof Promise)
+			return state
+
+		lifecycleReturn(element, state)
+	} catch (e) {
+		errorBoundary(element, e, name, 1)
+	}
+}
+
+/**
+ * @param {Element} element
+ * @param {string} name
+ * @param {Object} props
+ * @param {Object} state
+ * @param {Object} context
+ */
+function lifecycleUpdate (element, name, props, state, context) {
+	try {
+		var state = element.owner[name].call(element.instance, props, state, context)
+
+		if (typeof state !== 'object')
```

Comment:
why differ from above `state instanceof Promise`?

Reply:
`state instanceof Promise` is used in `lifecycleMount` since that handles `componentWillUnmount` and that supports a Promise as a return value for delayed unmounting. Here we want to return early non object values like booleans since `shouldComponentUpdate` expects this.

Reply:
In fact, i think we just use the `name` variable here and in `lifecycleMount ` to know where this is coming from and return the appropriate value.